OrbitGl/TimeGraph.cpp

```diff
@@ -764,364 +757,12 @@ void TimeGraph::DrawTracks(GlCanvas* canvas, PickingMode picking_mode) {
-      sorted_filtered_tracks_.push_back(moving_track);
-      moving_track_current_position = static_cast<int>(sorted_filtered_tracks_.size()) - 1;
-    }
-
-    // Now we have to change the position of the moving_track in the non-filtered array
-    if (moving_track_current_position == moving_track_previous_position) {
-      return;
-    }
-    sorted_tracks_.erase(std::find(sorted_tracks_.begin(), sorted_tracks_.end(), moving_track));
-    if (moving_track_current_position > moving_track_previous_position) {
-      // In this case we will insert the moving_track right after the one who is before in the
-      // filtered array
-      auto& previous_filtered_track = sorted_filtered_tracks_[moving_track_current_position - 1];
-      sorted_tracks_.insert(
-          ++std::find(sorted_tracks_.begin(), sorted_tracks_.end(), previous_filtered_track),
-          moving_track);
-    } else {
-      // In this case we will insert the moving_track right before the one who is after in the
-      // filtered array
-      auto& next_filtered_track = sorted_filtered_tracks_[moving_track_current_position + 1];
-      sorted_tracks_.insert(
-          std::find(sorted_tracks_.begin(), sorted_tracks_.end(), next_filtered_track),
-          moving_track);
-    }
-  }
-}
-
-int TimeGraph::FindMovingTrackIndex() {
-  // Returns the position of the moving track, or -1 if there is none.
-  for (auto track_it = sorted_filtered_tracks_.begin(); track_it != sorted_filtered_tracks_.end();
-       ++track_it) {
-    if ((*track_it)->IsMoving()) {
-      return track_it - sorted_filtered_tracks_.begin();
-    }
-  }
-  return -1;
-}
-
-void TimeGraph::UpdateTracks(uint64_t min_tick, uint64_t max_tick, PickingMode picking_mode) {
-  // Make sure track tab fits in the viewport.
-  float current_y = -layout_.GetSchedulerTrackOffset() - layout_.GetTrackTabHeight();
-  float pinned_tracks_height = 0.f;
-
-  // Draw pinned tracks
-  for (auto& track : sorted_filtered_tracks_) {
-    if (!track->IsPinned()) {
-      continue;
-    }
-
-    const float z_offset = GlCanvas::kZOffsetPinnedTrack;
-    track->SetY(current_y + canvas_->GetWorldTopLeftY() - layout_.GetTopMargin() -
-                layout_.GetSchedulerTrackOffset());
-    track->UpdatePrimitives(min_tick, max_tick, picking_mode, z_offset);
-    const float height = (track->GetHeight() + layout_.GetSpaceBetweenTracks());
-    current_y -= height;
-    pinned_tracks_height += height;
-  }
-
-  // Draw unpinned tracks
-  for (auto& track : sorted_filtered_tracks_) {
-    if (track->IsPinned()) {
-      continue;
-    }
-
-    const float z_offset = track->IsMoving() ? GlCanvas::kZOffsetMovingTack : 0.f;
-    track->SetY(current_y);
-    track->UpdatePrimitives(min_tick, max_tick, picking_mode, z_offset);
-    current_y -= (track->GetHeight() + layout_.GetSpaceBetweenTracks());
-  }
-
-  min_y_ = current_y;
-}
+void TimeGraph::SortTracks() { track_manager_->SortTracks(); }
```

Comment:
I liked this proxy methods to have a better defined parent-child relationship. 

App -> TimeGraph -> TrackManager -> Tracks
Why App has to know about TrackManager?

But, I'm completely open to discuss about that.

Reply:
Especially since the track manager is already exposed, I see three issues with this:
1) It increases the amount of methods in TimeGraph's API, making it less clear what's the scope of responsibility for this class.
2) Every change done in TrackManager needs to "bubble up". Just renaming methods or changing parameters requires changing this in TimeGraph as well
3) And for me, this is the most important one: It is not clear to the user which method to use, because both "TimeGraph::SortTracks()" and "TimeGraph::GetTrackManager()::SortTracks()" are available to them. You need to look at the implementation to figure out that those actually do the exact same thing.

Reply:
So, you recommend to do a GetTimeGraph->GetTrackManager -> SortTracks on App?

Reply:
If we keep the call to SortTracks in App, yes, I would suggest that.

But this also relates to what I called out regarding the name of `SortTracks`. We're only using this method in `OrbitApp` to update the sorting of the tracks. Why is this needed? Why does `OrbitApp` need to know that the `TimeGraph` wants to do a re-sorting of tracks after timer updates? Sorting is the domain of the `TrackManager`, and it should be responsible to decide when it's time to re-sort the tracks, not the App. **Especially** since the `TrackManager` already decides on its own if it actually wants to perform the sorting as instructed, or just does an early-out in `SortTracks` based on conditions not visible to the user of the class. 

The track manager should know when timers are added to the Tracks, and it can then trigger re-sorting of the tracks when it thinks it's required, but the App should not know nor care about this. 

Since currently timers are added through the `TimeGraph`, it's a reasonable compromise for now to have the `TimeGraph` decide when it's time to sort the tracks, but I'd argue that the Timers should be passed to `TrackManager`, then this class decides which tracks to add them to and if it's time to perform a sorting, and neither the App nor the TimeGraph need to care.

Reply:
90% agree. I'm ok (or at least not 100% sure of moving them) with Timers added from TimeGraph, but TrackManager should manage their own sorting of Tracks. 

Erased the method in TimeGraph and added a TODO for that.

Reply:
b/176077097